kiwmi/desktop/view.c

```diff
@@ -259,3 +299,229 @@ view_create(
+{
+    struct kiwmi_view_child *child = wl_container_of(listener, child, commit);
+    if (child->mapped && child->parent_mapped) {
+        view_child_damage(child);
+    }
+}
+
+static void
+view_child_new_subsurface_notify(struct wl_listener *listener, void *data)
+{
+    struct kiwmi_view_child *child =
+        wl_container_of(listener, child, new_subsurface);
+    struct wlr_subsurface *subsurface = data;
+    view_child_subsurface_create(child, child->view, subsurface);
+}
+
+static void
+view_child_map_notify(struct wl_listener *listener, void *UNUSED(data))
+{
+    struct kiwmi_view_child *child = wl_container_of(listener, child, map);
+    child->mapped                  = true;
+    if (child->parent_mapped) {
+        view_child_damage(child);
+        wl_signal_emit(&child->events.map, child);
+    }
+}
+
+static void
+view_child_unmap_notify(struct wl_listener *listener, void *UNUSED(data))
+{
+    struct kiwmi_view_child *child = wl_container_of(listener, child, unmap);
+    child->mapped                  = false;
+    if (child->parent_mapped) {
+        view_child_damage(child);
+        wl_signal_emit(&child->events.unmap, child);
+    }
+}
+
+static void
+view_child_parent_map_notify(struct wl_listener *listener, void *UNUSED(data))
+{
+    struct kiwmi_view_child *child =
+        wl_container_of(listener, child, parent_map);
+    if (child->mapped) {
+        child->parent_mapped = true;
+        view_child_damage(child);
+        wl_signal_emit(&child->events.map, child);
+    }
+}
+
+static void
+view_child_parent_unmap_notify(struct wl_listener *listener, void *UNUSED(data))
+{
+    struct kiwmi_view_child *child =
+        wl_container_of(listener, child, parent_unmap);
+    if (child->mapped) { // TODO: && child->parent_mapped? (also in map_notify)
+        child->parent_mapped = false;
+        view_child_damage(child);
+        wl_signal_emit(&child->events.unmap, child);
+    }
+}
+
+struct kiwmi_view_child *
+view_child_create(
+    struct kiwmi_view_child *parent,
+    struct kiwmi_view *view,
+    struct wlr_surface *wlr_surface,
+    enum kiwmi_view_child_type type,
+    const struct kiwmi_view_child_impl *impl)
+{
+    struct kiwmi_view_child *child = calloc(1, sizeof(*child));
+    if (!child) {
+        wlr_log(WLR_ERROR, "Failed to allocate view_child");
+        return NULL;
+    }
+
+    child->type        = type;
+    child->impl        = impl;
+    child->view        = view;
+    child->wlr_surface = wlr_surface;
+    child->mapped      = false;
+    // Q: does it make sense to cache it all the time just to avoid the invalid
+    // child->view pointer? Any better idea?
+    // We could destroy all children once they lose connection to their view
+    // (i.e. the view or their parent is destroyed), because if the surfaces are
+    // reused, we create new kiwmi_view_child'ren anyway.
+    child->parent_mapped =
+        parent ? parent->mapped && parent->parent_mapped : view->mapped;
```

Comment:
I was just lazy to think it through at the time of writing. That said, comments are welcome, but maybe I can sort it out myself.

Reply:
If they get recreated anyways, just destroy them I'd say. Am I missing something?

Reply:
I was just unsure if I saw everything.